### PR TITLE
Renesas R-Car: Add PWM driver

### PR DESCRIPTION
**Pull request by pmarzin**
This PR adds the PWM driver to support the PWM controller of R-Car SoC family.

The dts has been kept as much close as Linux as possible.
Unlike the dts, the driver is slightly different as the Zephyr pwm api differs.
Clock divider is updated only when possible.

Actually, the api driver test for the pwm (tests/drivers/pwm/pwm_api/src/test_pwm.c) does not work on R-Car as the test is using prohibited values (0% of duty cycle and cycle value > 1023 are not supported by the R-Car PWM controller).
Nevertheless, it has been tested on H3ULCB with the blinky pwm sample with success.

[UPDATE]
For the pwm_api test, a proposal is to adapt the test in order to not test pwm API with prohibited values. See the related PR here : https://github.com/zephyrproject-rtos/zephyr/pull/48834

### CODE REVIEW

File: boards/arm/rcar_h3ulcb/rcar_h3ulcb_cr7-pinctrl.dtsi
Comment (julien-massot) [inline boards/arm/rcar_h3ulcb/rcar_h3ulcb_cr7-pinctrl.dtsi:34]:
```suggestion
	pwm0_default: pwm0_default {
```

File: drivers/pwm/pwm_rcar.c
Comment (julien-massot) [inline drivers/pwm/pwm_rcar.c:23]: Please drop these helper macros and use dev->data, dev->config directly

File: drivers/pwm/pwm_rcar.c
Comment (julien-massot) [inline drivers/pwm/pwm_rcar.c:140]: perhaps add a comment here, that on R-Car we only support one gpio per instance, each of them are declared individually to match Linux dts .

File: drivers/pwm/pwm_rcar.c
Comment (julien-massot) [inline drivers/pwm/pwm_rcar.c:136]: For this one and all the following
```suggestion
	const struct pwm_rcar_cfg *config = dev->config;
```

File: drivers/pwm/pwm_rcar.c
Comment (julien-massot) [inline drivers/pwm/pwm_rcar.c:204]:
```suggestion
	const struct pwm_rcar_cfg *config = dev->config;
	struct pwm_rcar_data *data = dev->data;
```

File: dts/arm/renesas/gen3/rcar_gen3_cr7.dtsi
Comment (julien-massot) [inline dts/arm/renesas/gen3/rcar_gen3_cr7.dtsi:114]: "This LSI incorporates seven channels for R-Car H3, R-Car H3-N, R-Car M3-W, R-Car M3-N, R-Car E3; five channels
for R-Car V3M and R-Car V3H; four channels for R-Car D3 pulse width modulation (PWM) timer."

Since all Gen3 SoCs have different number of pwm channels we may move these pwm definition in the SoC specific dts "r8a77951.dtsi".

File: dts/bindings/pwm/renesas,rcar-pwm.yaml
Comment (julien-massot) [inline dts/bindings/pwm/renesas,rcar-pwm.yaml:6]: just noticed that Linux is using 
"renesas,pwm-rcar"
https://elixir.bootlin.com/linux/v5.17-rc3/source/Documentation/devicetree/bindings/pwm/renesas,pwm-rcar.yaml
Please renease this file and the compatible entry.

File: dts/arm/renesas/gen3/rcar_gen3_cr7.dtsi
Comment (julien-massot) [inline dts/arm/renesas/gen3/rcar_gen3_cr7.dtsi:82]: Nits, Linux is describing it in this order:
- compatible
- reg
- clocks
- pwm-cells
- status

Comment (julien-massot) [reply dts/arm/renesas/gen3/rcar_gen3_cr7.dtsi:82]: https://github.com/torvalds/linux/blob/master/arch/arm64/boot/dts/renesas/r8a77951.dtsi

File: drivers/pwm/pwm_rcar.c
Comment (julien-massot) [inline drivers/pwm/pwm_rcar.c:160]: Does by just disabling the pwm you can simulate the 0% duty cycle ?

Comment (pmarzin) [reply drivers/pwm/pwm_rcar.c:160]: No, nevertheless, disabling the pwm will set a high level on the channel pin. Like a duty cycle at 100%.

File: boards/arm/rcar_h3ulcb/rcar_h3ulcb_cr7-pinctrl.dtsi
Comment (pmarzin) [reply boards/arm/rcar_h3ulcb/rcar_h3ulcb_cr7-pinctrl.dtsi]: Thanks, this has been corrected.

File: dts/bindings/pwm/renesas,rcar-pwm.yaml
Comment (pmarzin) [reply dts/bindings/pwm/renesas,rcar-pwm.yaml]: Corrected. Now declared like Linux dts files.

File: dts/arm/renesas/gen3/rcar_gen3_cr7.dtsi
Comment (pmarzin) [reply dts/arm/renesas/gen3/rcar_gen3_cr7.dtsi:82]: Corrected, thanks. Now follows the Linux pwm-rcar organization.

File: drivers/pwm/pwm_rcar.c
Comment (pmarzin) [reply drivers/pwm/pwm_rcar.c]: Done :)

File: drivers/pwm/pwm_rcar.c
Comment (pmarzin) [reply drivers/pwm/pwm_rcar.c:137]: Yes indeed, to be the closest as Linux as possible (regarding the dts), only one pwm channel is allowed per instance. I put a comment. Thanks.

File: drivers/pwm/pwm_rcar.c
Comment (julien-massot) [inline drivers/pwm/pwm_rcar.c:152]: (config->reg_addr & 0x0000f000) >> 3 ? 

What was the intention ?

Comment (pmarzin) [reply drivers/pwm/pwm_rcar.c:152]: According to Table 60.2 from https://www.renesas.com/us/en/document/man/r-car-h3-m3-users-manual documentation, this is to get the PWM channel number, extracted from the reg_addr value.

**Review by mbolivar-nordic — DISMISSED** [review]
Devicetree portions LGTM; nice job. I did not review the rest of the driver.

**Review by julien-massot — DISMISSED** [review]

File: drivers/pwm/pwm_rcar.c
Comment (pmarzin) [reply drivers/pwm/pwm_rcar.c]: To simplify, this has been modified. Now, the base address register is printed. Base register will help to identify which PWM channel is used.

**Review by julien-massot — DISMISSED** [review]

File: boards/arm/rcar_h3ulcb/doc/rcar_h3ulcb.rst
Comment (aaillet) [inline boards/arm/rcar_h3ulcb/doc/rcar_h3ulcb.rst:176]:
```suggestion
When plugged on a Kingfisher daughter board, pwm4 channel is available on CN7 LVDS connector.
```

File: boards/arm/rcar_h3ulcb/rcar_h3ulcb_cr7.yaml
Comment (aaillet) [inline boards/arm/rcar_h3ulcb/rcar_h3ulcb_cr7.yaml:14]: If you do not support the pwm test suite (according to your PR message), you should not mark it as supported in your board's yaml file. Otherwise, you will get errors on the pwm suite every time you run twister on your board.
So I think that you should mark this suite as ignored for now before marking it as supported later if you find a way to patch either your driver or the test suite.

File: drivers/pwm/pwm_rcar.c
Comment (aaillet) [inline drivers/pwm/pwm_rcar.c:152]: ENOTSUP as it is not supported by the pwn controller ?

File: drivers/pwm/pwm_rcar.c
Comment (aaillet) [inline drivers/pwm/pwm_rcar.c:180]: Spaces and tabulation are mixed here, please unify this in the driver file.

File: drivers/pwm/pwm_rcar.c
Comment (aaillet) [inline drivers/pwm/pwm_rcar.c:246]: `int ret = 0;` then `return ret;` ?

**Review by aaillet — CHANGES_REQUESTED** [review]

File: drivers/pwm/pwm_rcar.c
Comment (pmarzin) [reply drivers/pwm/pwm_rcar.c:180]: Theses spaces are made by uncrustify to align the "if" arguments.

File: boards/arm/rcar_h3ulcb/doc/rcar_h3ulcb.rst
Comment (pmarzin) [reply boards/arm/rcar_h3ulcb/doc/rcar_h3ulcb.rst:176]: Your proposal is easier to understand. Agreed.

File: boards/arm/rcar_h3ulcb/rcar_h3ulcb_cr7.yaml
Comment (pmarzin) [reply boards/arm/rcar_h3ulcb/rcar_h3ulcb_cr7.yaml:14]: Good point here. I will rework this until the test could be passed with success.

File: drivers/pwm/pwm_rcar.c
Comment (pmarzin) [reply drivers/pwm/pwm_rcar.c]: ret is always modified so no need to init it to 0.
I'm ok to simplify the end as `clock_control_get_rate` return 0 if no error.

File: drivers/pwm/pwm_rcar.c
Comment (pmarzin) [reply drivers/pwm/pwm_rcar.c]: 0 is a prohibited value so "Unsupported value" is preferred than "Invalid argument". Correct. Fixed. Thanks :)

**Review by julien-massot — APPROVED** [review]

**Review by aaillet — APPROVED** [review]

**Comment by pmarzin** [comment]
Hi @mbolivar-nordic ! Some basic improvements and rebase on main have been done (without modification on device tree parts). Can you update your review please ?

**Review by aaillet — DISMISSED** [review]

**Review by julien-massot — DISMISSED** [review]

**Comment by julien-massot** [comment]
@henrikbrixandersen would you like to check this PR ?

File: drivers/pwm/pwm_rcar.c
Comment (henrikbrixandersen) [inline drivers/pwm/pwm_rcar.c:149]:
```suggestion
	/* Prohibited values */
```

File: drivers/pwm/pwm_rcar.c
Comment (henrikbrixandersen) [inline drivers/pwm/pwm_rcar.c:152]:
```suggestion
		return -EINVAL;
```

**Review by henrikbrixandersen — CHANGES_REQUESTED** [review]

File: drivers/pwm/pwm_rcar.c
Comment (pmarzin) [reply drivers/pwm/pwm_rcar.c]: Fixed! Thanks for your feedback :)

File: drivers/pwm/pwm_rcar.c
Comment (pmarzin) [reply drivers/pwm/pwm_rcar.c]: Fixed! Thanks for your feedback :)

**Review by aaillet — DISMISSED** [review]

**Review by aaillet — DISMISSED** [review]

**Review by mbolivar-nordic — DISMISSED** [review]
Looks good to me.

File: dts/bindings/pwm/renesas,pwm-rcar.yaml
Comment (henrikbrixandersen) [inline dts/bindings/pwm/renesas,pwm-rcar.yaml:16]:
```suggestion
pwm-cells:
  - channel
# period in terms of nanoseconds
  - period
  - flags
```

This is the standard convention.

**Review by henrikbrixandersen — CHANGES_REQUESTED** [review]

File: dts/bindings/pwm/renesas,pwm-rcar.yaml
Comment (pmarzin) [reply dts/bindings/pwm/renesas,pwm-rcar.yaml:16]: Hi Henrik, thanks. I didn't saw this conventional. I will update it based on the #45013 PR.

Comment (pmarzin) [reply dts/bindings/pwm/renesas,pwm-rcar.yaml]: Ok, Done ! I did the update with a rebase on main. Do not hesitate to tell me if you see anything else. Thank you @henrikbrixandersen

File: boards/arm/rcar_h3ulcb/rcar_h3ulcb_cr7.yaml
Comment (henrikbrixandersen) [inline boards/arm/rcar_h3ulcb/rcar_h3ulcb_cr7.yaml:17]: I just noticed you are ignoring all `pwm` tests/samples for this board - how come?

**Review by henrikbrixandersen — CHANGES_REQUESTED** [review]

File: boards/arm/rcar_h3ulcb/rcar_h3ulcb_cr7.yaml
Comment (pmarzin) [reply boards/arm/rcar_h3ulcb/rcar_h3ulcb_cr7.yaml:17]: As mentioned into the PR message, pwm_api tests use prohibited values that are not hardware supported by our PWM controller. That's why with @aaillet we decided to put it as "ignored" in order to not block twister test suite.
Should we propose a fix into the pwm_api test to add a particular case for this driver ? What is your advise ?
Thanks

Comment (henrikbrixandersen) [reply boards/arm/rcar_h3ulcb/rcar_h3ulcb_cr7.yaml:17]: > As mentioned into the PR message, pwm_api tests use prohibited values that are not hardware supported by our PWM controller. That's why with @aaillet we decided to put it as "ignored" in order to not block twister test suite.
> Should we propose a fix into the pwm_api test to add a particular case for this driver ? What is your advise ?

The common PWM tests should be driver agnostic. I'd submit an issue - or even a PR, if you have a proposal for a fix.

**Review by henrikbrixandersen — DISMISSED** [review]

**Review by aaillet — DISMISSED** [review]

**Comment by pmarzin** [comment]
The PR is currently blocked by failure tests that are out of PWM scope. They are known and followed by this PR #45102 . Waiting it to solve the tests suite.

**Comment by pmarzin** [comment]
As PR #45102 is merged, "update with rebase" has been triggered to solve tests issues. All are good now. This PR is ready :)
@henrikbrixandersen @aaillet  @mbolivar-nordic : sorry for the re-ask : Can you update your approvement please ?

File: drivers/pwm/pwm_rcar.c
Comment (henrikbrixandersen) [inline drivers/pwm/pwm_rcar.c:251]: You need to catch up with the recent changes to the PWM API (see 95b56cdffd61f2843b0533d7ab0f1f1ccc9dc092, b317ce8597d471cdc6df117a5cc0877b783f6af7, and related commits)

**Review by henrikbrixandersen — CHANGES_REQUESTED** [review]

File: drivers/pwm/pwm_rcar.c
Comment (pmarzin) [reply drivers/pwm/pwm_rcar.c]: This PR has been updated with the mentioned commits and some others like https://github.com/zephyrproject-rtos/zephyr/commit/885087b3288346e2e09da92e506e5f390cd4ad29.

**Review by henrikbrixandersen — DISMISSED** [review]
LGTM

File: boards/arm/rcar_h3ulcb/rcar_h3ulcb_cr7.yaml
Comment (aaillet) [inline boards/arm/rcar_h3ulcb/rcar_h3ulcb_cr7.yaml:14]: Should maybe set to unsupported waiting for #48834 to be merged ? Twister suite will report errors in this state !
What do you think @henrikbrixandersen, it depend on Zephyr test suite politic, maybe you are more aware than me ?

Comment (henrikbrixandersen) [reply boards/arm/rcar_h3ulcb/rcar_h3ulcb_cr7.yaml:14]: Either that or exclude the board from the tests for now.

Comment (pmarzin) [reply boards/arm/rcar_h3ulcb/rcar_h3ulcb_cr7.yaml]: Yes indeed, otherwise the twister test will failed. I removed it !
I did a push force again but everything else remains intact :)

Comment (aaillet) [reply boards/arm/rcar_h3ulcb/rcar_h3ulcb_cr7.yaml]: Thanks

**Review by aaillet — DISMISSED** [review]

**Review by henrikbrixandersen — DISMISSED** [review]

**Comment by pmarzin** [comment]
Sorry, due to a failure test I had to remove "label" property from the dts file.

**Review by henrikbrixandersen — APPROVED** [review]

**Review by aaillet — APPROVED** [review]

**Comment by carlescufi** [comment]
Merging this without @anangl's approval due to delays in reviews.

**Comment by pmarzin** [comment]
🥳 Thanks for your help guys 👍🏼

**Pull request merged**
